src/promotion/nhit/hash.c

```diff
@@ -0,0 +1,393 @@
+ *	+--++-+--+---+-+--+--+--++-+--+--++-+--++-+
+ *	|  |  |  |  |X |  |  |  |  |  |  |  |  |  | ring_buffer
+ *	+--++-+--+---+-+--+--+--++-+--+--++-+--+--+
+ *	    |        ^           |        ^
+ *	    |________|           |________|
+ *
+ * Since rb_pointer is pointing to occupied rb slot we need to write-lock hash
+ * bucket I associated with this slot and remove it from collision list.
+ * We've gained an empty slot and we use slot X for new hash H entry.
+ *
+ *	+--+--+--+--+--+--+--+--+--+--+
+ *	|  |  |I |  |  |  |  |  |H |  | hash_map
+ *	+--+--++-+--+--+--+--+--++-+--+
+ *	     __|   rb_pointer    |        _______
+ *	    |           +        |        |     |
+ *	    v           v        v        |     v
+ *	+--++-+--+-----++-+--+--++-+--+--++-+--++-+
+ *	|  |  |  |  |X |  |  |  |  |  |  |  |  |  | ring_buffer
+ *	+--+--+--+---+-+--+--+--++-+--+--++-+--++-+
+ *		     ^           |        ^     |
+ *		     |           |________|     |
+ *		     |__________________________|
+ *
+ * Valid field in nhit_list_elem is guarded by rb_pointer_lock to make sure we
+ * won't try to use the same slot in two threads. That would be possible if in
+ * time between removal from collision and insertion into the new one the
+ * rb_pointer would go around the whole structure (likeliness depends on size of
+ * ring_buffer).
+ */
+
+#define HASH_PRIME 4099
+
+struct nhit_list_elem {
+	ocf_core_id_t core_id;
+	uint64_t core_lba;
+
+	ocf_cache_line_t coll_prev;
+	ocf_cache_line_t coll_next;
+
+	bool valid;
+
+	env_atomic counter;
+};
+
+struct nhit_hash {
+	ocf_cache_line_t hash_entries;
+	uint64_t rb_entries;
+
+	ocf_cache_line_t *hash_map;
+	env_rwsem *hash_locks;
+
+	struct nhit_list_elem *ring_buffer;
+	uint64_t rb_pointer;
+	env_spinlock rb_pointer_lock;
+};
+
+ocf_error_t hash_init(uint64_t hash_size, nhit_hash_t *ctx)
+{
+	int result = 0;
+	struct nhit_hash *new_ctx;
+	uint64_t i;
+
+	new_ctx = env_vzalloc(sizeof(*new_ctx));
+	if (!new_ctx) {
+		result = -OCF_ERR_NO_MEM;
+		goto exit;
+	}
+
+	new_ctx->rb_entries = hash_size;
+	new_ctx->hash_entries = OCF_DIV_ROUND_UP(
+			new_ctx->rb_entries / 4,
+			HASH_PRIME) * HASH_PRIME - 1;
+
+	new_ctx->hash_map = env_vzalloc(
+			new_ctx->hash_entries * sizeof(*new_ctx->hash_map));
+	if (!new_ctx->hash_map) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_ctx;
+	}
+	for (i = 0; i < new_ctx->hash_entries; i++)
+		new_ctx->hash_map[i] = new_ctx->rb_entries;
+
+	new_ctx->hash_locks = env_vzalloc(
+			new_ctx->hash_entries * sizeof(*new_ctx->hash_locks));
+	if (!new_ctx->hash_locks) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_hash;
+	}
+
+	for (i = 0; i < new_ctx->hash_entries; i++) {
+		if (env_rwsem_init(&new_ctx->hash_locks[i])) {
+			result = -OCF_ERR_UNKNOWN;
+			goto dealloc_locks;
+		}
+	}
+
+	new_ctx->ring_buffer = env_vzalloc(
+			new_ctx->rb_entries * sizeof(*new_ctx->ring_buffer));
+	if (!new_ctx->ring_buffer) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_locks;
```

Comment:
You should probably add deinit_rwsem label here ...

Reply:
Got some more patches lined up after this PR, will fix it there!

Reply:
Got some more patches lined up after this PR, will fix it there!